File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/SuggestionsAreMade.cs
namespace SeatsSuggestions;

/// <summary>
///     Occurs when a bunch of Suggestion are made.
/// </summary>
public class SuggestionsAreMade
{
    public SuggestionsAreMade(ShowId showId, PartyRequested partyRequested)
    {
        ShowId = showId;
        PartyRequested = partyRequested;
        InstantiateAnEmptyListForEveryPricingCategory();
    }

    public ShowId ShowId { get; }
    public PartyRequested PartyRequested { get; }

    public Dictionary<PricingCategory, List<SuggestionIsMade>> ForCategory { get; } = new();

    public IEnumerable<string> SeatNames(PricingCategory pricingCategory)
    {
        var seatingPlaces = ForCategory[pricingCategory]
            .SelectMany(s => s.SuggestedSeats)
            .ToList();

        return seatingPlaces.Select(s => s.ToString());
    }

    private void InstantiateAnEmptyListForEveryPricingCategory()
    {
        foreach (PricingCategory pricingCategory in Enum.GetValues(typeof(PricingCategory)))
            ForCategory[pricingCategory] = [];
    }

    public void Add(IEnumerable<SuggestionIsMade> suggestions)
    {
        foreach (var suggestionMade in suggestions) ForCategory[suggestionMade.PricingCategory].Add(suggestionMade);
    }

    public bool MatchExpectations()
    {
        return ForCategory.SelectMany(s => s.Value).Any(x => x.MatchExpectation());
    }
}

File: TheaterSuggestions/Java/SeatingSuggestions.Domain/src/main/java/org/octo/SeatingPlaceSuggestions/Domain/SuggestionsAreMade.java
package org.octo.SeatingPlaceSuggestions.Domain;

import java.util.*;

public class SuggestionsAreMade {
    private final ShowID showId;
    private final PartyRequested partyRequested;

    private final Map<PricingCategory, List<SuggestionIsMade>> forCategory = new HashMap<>();

    public SuggestionsAreMade(ShowID showId, PartyRequested partyRequested) {
        this.showId = showId;
        this.partyRequested = partyRequested;

        instantiateAnEmptyListForEveryPricingCategory();
    }

    public List<String> seatNames(PricingCategory pricingCategory) {
        return forCategory.get(pricingCategory).stream()
                .map(SuggestionIsMade::seatNames).flatMap(Collection::stream).toList();
    }

    private void instantiateAnEmptyListForEveryPricingCategory() {
        for (PricingCategory pricingCategory : PricingCategory.values()) {
            forCategory.put(pricingCategory, new ArrayList<>());
        }
    }

    public void add(Iterable<SuggestionIsMade> suggestions) {
        suggestions.forEach(suggestionIsMade -> forCategory.get(suggestionIsMade.pricingCategory()).add(suggestionIsMade));
    }

    public boolean matchExpectations() {
        return forCategory.values().stream().flatMap(List::stream).anyMatch(SuggestionIsMade::matchExpectation);
    }

    public ShowID showId() {
        return showId;
    }

    public PartyRequested partyRequested() {
        return partyRequested;
    }
}
